stout/composed-loop.h

```diff
@@ -0,0 +1,154 @@
+#pragma once
+
+#include <optional>
+
+#include "stout/stream.h"
+#include "stout/terminal.h"
+
+////////////////////////////////////////////////////////////////////////
+
+namespace stout {
+namespace eventuals {
+
+////////////////////////////////////////////////////////////////////////
+
+using namespace detail;
+
+struct _ComposedLoopRouter {
+  template <typename PrevK_, typename Arg_>
+  struct Continuation {
+    void Start(TypeErasedStream& stream) {
+      stream.Next();
+    }
+
+    template <typename... Args>
+    void Body(Args&&... args) {
+      stored_k_->k_.Body(std::forward<Args>(args)...);
+    }
+
+    void Ended() {
+      stored_k_->Done();
+    }
+
+    PrevK_* stored_k_;
+  };
+
+  template <typename ComposedK>
+  struct Composable {
+    template <typename Arg>
+    using ValueFrom = Arg;
+
+    template <typename Arg>
+    auto k() && {
+      return Continuation<ComposedK, Arg>{std::move(composed_k)};
+    }
+
+    ComposedK* composed_k;
+  };
+};
+
+template <typename K>
+auto ComposedLoopRouter(K* k) {
+  return _ComposedLoopRouter::Composable<K>{k};
+}
+
+////////////////////////////////////////////////////////////////////////
+
+struct _ComposedLoop {
+  template <typename K_, typename F_, typename Arg_>
+  struct Continuation : public TypeErasedStream {
+    // NOTE: explicit constructor because inheriting 'TypeErasedStream'.
+    Continuation(K_ k, F_ f)
+      : k_(std::move(k)),
+        f_(std::move(f)) {}
+
+    void Start(TypeErasedStream& stream) {
+      stream_ = &stream;
+      k_.Start(*this);
+    }
+
+    template <typename... Args>
+    void Fail(Args&&... args) {
+      k_.Fail(std::forward<Args>(args)...);
+    }
+
+    void Stop() {
+      k_.Stop();
+    }
+
+    template <typename... Args>
+    void Body(Args&&... args) {
+      // Executes when Body is called from the above-level stream.
+      if (innerStream_ == nullptr) {
+        inner_.emplace(Build(operator|(f_(std::forward<Args>(args)...), ComposedLoopRouter(this))));
```

Comment:
Our formatting check doesn't enforce the 80 character rule yet, so please manually check that all the lines are <80 characters.